applications/vanilla/models/class.categorymodel.php

```diff
@@ -428,6 +439,66 @@ public static function categoryWatch($honorHideAllDiscussion = true) {
         return $watch;
     }
 
+    /**
+     * Get a list of IDs of categories visible to the current user.
+     *
+     * @param array $options
+     * @return array|bool An array of filtered categories or true if no categories were filtered.
+     */
+    public function getVisibleCategories(array $options = []) {
+        $categories = self::categories();
+        $unfiltered = true;
+        $result = [];
+
+        // Options
+        $filterHideDiscussions = $options['filterHideDiscussions'] ?? false;
+
+        foreach ($categories as $categoryID => $category) {
+            if ($filterHideDiscussions && val('HideAllDiscussions', $category)) {
+                $unfiltered = false;
+                continue;
+            }
+
+            if ($category['PermsDiscussionsView']) {
+                $result[] = $category;
+            } elseif ($unfiltered) {
+                $unfiltered = false;
+            }
+        }
+
+        if ($unfiltered) {
+            $result = true;
+        }
+
+        // Allow addons to modify the visible categories.
+        $result = $this->eventManager->fireFilter('categoryModel_visibleCategories', $result);
+
+        return $result;
+    }
+
+    /**
+     * Get a list of IDs of categories visible to the current user.
+     *
+     * @see CategoryModel::categoryWatch
+     * @param array $options
+     * @return array|bool An array of filtered category IDs or true if no categories were filtered.
+     */
+    public static function getVisibleCategoryIDs(array $options = []) {
```

Comment:
We should probably make that a non static method and force the caller to use `CategoryModel::instance()-> getVisibleCategoryIDs()`